### PR TITLE
config: prefixes image names with ko:// scheme

### PR DESCRIPTION
**Pull request by poy**
# Changes

Prefixes image names with ko:// scheme. This allows `ko` to improve the experience with Go modules. See https://github.com/google/ko/pull/58 for more information.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

### CODE REVIEW

File: config/controller.yaml [hunk 1]
```diff
@@ -42,7 +42,7 @@ spec:
       serviceAccountName: tekton-pipelines-controller
       containers:
       - name: tekton-pipelines-controller
-        image: github.com/tektoncd/pipeline/cmd/controller
+        image: ko://github.com/tektoncd/pipeline/cmd/controller
         args: [
           "-kubeconfig-writer-image", "github.com/tektoncd/pipeline/cmd/kubeconfigwriter",
```
Comment (vdemeester) [inline config/controller.yaml]: Shouldn't thoses (all starting with `github.com/tektoncd/pipeline`) also be prefixed ?

**Comment by ImJasonH** [comment]
/ok-to-test

The improvement is from making typos less difficult to debug.

**Comment by vdemeester** [comment]
@ImJasonH could there be a `strict` way with this, as : in a "strict mode" you would get an error if a `ko://` reference is missing from what `ko` gathered to build the image for. Would it be possible to have that ? (because otherwise, it will fail silently, right ? and will have an ErrPull… on the pod(s))

**Comment by ImJasonH** [comment]
If I'm understanding, yes that's exactly what ko does with this prefix. If it sees the special prefix in normal mode it strips it (for recent versions of ko in the last 6? months). If you pass `--strict` then it strips the ko:// prefix _and_ fails if the import path isn't recognized, instead of passing it through, which leads to unpullable image references.

This is step one to adopting ko strict more. After this we can run ko with --strict to avoid bad refs.

**Comment by vdemeester** [comment]
> If I'm understanding, yes that's exactly what ko does with this prefix. If it sees the special prefix in normal mode it strips it (for recent versions of ko in the last 6? months). If you pass `--strict` then it strips the ko:// prefix _and_ fails if the import path isn't recognized, instead of passing it through, which leads to unpullable image references.
> 
> This is step one to adopting ko strict more. After this we can run ko with --strict to avoid bad refs.

Alright, if it works like this, it sounds good to me :+1: 
/lgtm

/cc @afrittoli @bobcatfish

**Comment by vdemeester** [comment]
https://github.com/tektoncd/pipeline/pull/2216/files/f7442117ba4f54a884173bebb06ff542e7bc6ffe#diff-d56e306ce931c2991f18ef68b6b501a3 needs to be taken into account :angel:

**Comment by vdemeester** [comment]
/hold cancel
I'll follow-up on this

**Pull request merged**